components/signoffs/signoffController.js

```diff
@@ -0,0 +1,170 @@
+
+const getSignoff = async (req, res, next) => {
+  try {
+    const signoffs = await signoffDao.getByCaseIdentifier(
+      req.params.caseIdentifier
+    );
+    if (signoffs) {
+      res.status(200).json(signoffs);
+    } else {
+      res.status(404).end();
+    }
+  } catch (e) {
+    handleErrors(e, 'Error getting signoff(s) for case', logger, next);
+  }
+};
+
+const addSignoff = async (req, res, next) => {
+  try {
+    const existingSignoff = null;
+    //const existingSignoff = await signoffDao.getByCaseIdentifier(
+    //  req.body.caseIdentifier
+    //);
+    //const validationResults = validateObjectsFromUser(req.body);
+    if (existingSignoff == null) {
+      await upsert(req.body);
+      res.status(201).end();
+    } else if (
+      existingSignoff.username == req.body.username &&
+      existingSignoff.signoff_step_name == req.body.signoff_step_name &&
+      existingSignoff.deliverable_type == req.body.deliverable_type
+    ) {
+      // signoff step is same, replace the old signoff record
+      await upsert(req.body);
+      res.status(201).end();
+    } else {
+      // signoff data is different
+      await upsert(req.body);
+      res.status(201).end();
+    }
+  } catch (e) {
+    handleErrors(e, 'Error adding cases', logger, next);
+  }
+};
+
+const upsert = (caseInfo) => {
+  return signoffDao.addSignoff(caseInfo);
+};
+
+const deleteSignoff = async (req, res, next) => {
+  try {
+    if (!req.body.id)
+      throw Error(400, 'Error: no "signoffId" provided request body');
+
+    const username = validateUsername(req.body.username);
+
+    const result = await signoffDao.deleteSignoff(req.body.id);
+    res.status(200).json(result);
+    next();
+  } catch (e) {
+    handleErrors(e, 'Error deleting records', logger, next);
+  }
+};
+
+function validateUsername (param) {
+  const user = nullifyIfBlank(param);
+  if (user == null || !user.length)
+    return new ValidationError('username must be provided');
+  if (user.match(/\W+/))
+    return new ValidationError('username must contain only letters');
+  return user;
+}
+
+function validateComment (param) {
+  let comment = nullifyIfBlank(param);
+  if (comment !== null)
+    comment = decodeURIComponent(comment.replace(/\+/g, ' '));
+  return comment;
+}
+
+function nullifyIfBlank (value) {
+  if (typeof value == 'undefined' || value === null || value.length == 0)
+    value = null;
+  return value;
+}
+
+function validateStepName (param) {
+  let stepname = nullifyIfBlank(param); //required by endpoint so shouldn't ever nullify
+  if (stepname !== 'undefined' && stepname !== null && stepname.length) {
+    stepname = stepname.toUpperCase();
+  }
+  let validSteps = [
+    'CASE REVIEW',
+    'ANALYSIS REVIEW',
+    'RELEASE APPROVAL',
+    'RELEASED',
+  ];
+  if (!validSteps.includes(stepname)) {
+    return new ValidationError(
+      'Sign-off must be associated with a valid step name: "CASE REVIEW", "ANALYSIS REVIEW", "RELEASE APPROVAL", or "RELEASED"'
```

Comment:
You might consider combining the base error message with the string contents of `validSteps` to create the full error message, rather than having to manually synchronize `validSteps` with the error message contents.